site/en/docs/extensions/mv3/automated-testing/index.md

```diff
@@ -0,0 +1,149 @@
+---
+layout: "layouts/doc-post.njk"
+title: "Automated testing for Chrome Extensions"
+seoTitle: "Automated testing for Chrome Extensions"
+date: 2023-10-09
+description: How to write automated tests for extensions.
+---
+
+Automated testing tools provide a great way to regularly test a large amount of your extension's
+functionality. This can be useful to ensure key functionality is working and as a way to avoid
+regressions between releases.
+
+## Unit testing
+
+[Unit testing][unit-testing] allows small sections of code to be tested in isolation from the rest
+of your extension, and outside of the browser.
+
+Code written without using extension APIs can be tested as normal, using a framework such as
+[Jest][jest]. To make code easier to test this way, consider using techniques such as
+[dependency injection][dependency-injection] which can help to remove dependencies on the chrome
+namespace in your lower level implementation.
+
+If you need to test code which includes extension APIs, consider using mocks.
+
+### Example: Using mocks with Jest
+
+Create a `jest.config.js` file, which declares a setup file that will run before all tests:
+
+```js
+module.exports = {
+  setupFiles: ['<rootDir>/mock-extension-apis.js']
+};
+```
+
+In `mock-extension-apis.js`, add implementations for the specific functions you expect to call:
+
+```js
+global.chrome = {
+  tabs: {
+    query: async () => { throw new Error("Unimplemented.") };
+  }
+};
+```
+
+Then, use jest.spy to mock a return value in a test:
+
+```js
+test("getActiveTabId returns active tab ID", async () => {
+  jest.spyOn(chrome.tabs, "query").mockResolvedValue([{
+    id: 3,
+    active: true,
+    currentWindow: true
+  }]);
+  expect(await getActiveTabId()).toBe(3);
+});
+```
+
+## Integration testing
+
+[Integration testing][integration-testing] involves an extension package being built and loaded into
+a browser. A testing tool communicates with the browser to automate interactions and test the same
+flows that a user would go through.
+
+See [Testing Chrome Extensions with Puppeteer][tutorial] for a tutorial.
+
+### Libraries
+
+Extensions are supported by a range of integration testing libraries.
+
+| Library                | Guidance                                                                                                                              |
+|------------------------|:-------------------------------------------------------------------------------------------------------------------------------------:|
+| Puppeteer / Playwright | See Chrome Extensions ([Puppeteer][puppeteer-testing] / [Playwright][playwright-testing]).                                            |
```

Comment:
Why are both in the same column?

Reply:
Playwright is a fork of Puppeteer and they generally work very similar (for example, if you compare https://pptr.dev/guides/chrome-extensions and https://playwright.dev/docs/chrome-extensions it is basically the same doc with the header changed).